### PR TITLE
feat: remove AsRunLog collection

### PR DESCRIPTION
**Pull request by Julusian**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

feature

* **What is the current behavior?** (You can also link to an open issue here)

Timings are stored in the AsRunLog collection, to be read by the `onAsRunEvent` blueprint method.
This works fine for shows a few hours long, but data in this collection is being retained for 50 days, and the blueprints often need to load all the AsRunLog for the current rundown (the log has no knowledge of activations). This can result in many tens to hundreds of thousands of documents in this collection which get loaded into memory.

While the events are retained for 50 days, none of the pieceInstances/partInstances/segments/rundown referenced by the events are, meaning the events will be hard to interpret without a snapshot or reingesting the rundown and some interpretation to rematch the ids

* **What is the new behavior (if this is a feature change)?**

We have always also stored the timings on the Parts & Pieces, but these could not be relied upon as they would be overwritten/deleted at various times. Since PartInstances, those timings are on the PartInstances and PieceInstances, and once set are never changed or removed until the Rundown is deleted from Sofie.
The `onAsRunEvent` blueprint method has been rethought to be able to operate directly on the PartInstance and PieceInstance collections for this data.

Additional improvements:
* Only find 'events' from the current activation
* Reduces DB operations, as the PartInstances would be loaded for some of the events
* Knowledge of segment 'playback', so that events can be found for the current play of a segment

As a consequence the 50 day retention of events has been removed. If this is desired again, then either the PartInstances and PieceInstances could be preserved, or snapshots could be auto-generated when removing a rundown.

* **Other information**:

**Status**

- [ ] Code documentation for the relevant parts in the code have been added/updated by the PR author
- [ ] The functionality has been tested by the PR author
- [ ] The functionality has been tested by NRK

### CODE REVIEW

File: packages/blueprints-integration/src/rundown.ts [hunk 1]
```diff
@@ -292,6 +299,13 @@ export interface IBlueprintPieceInstance<TMetadata = unknown> {
 
 	piece: IBlueprintPieceDB<TMetadata>
 
+	/** The time the system started playback of this part, null if not yet played back (milliseconds since epoch) */
+	startedPlayback?: Time
```
Comment (sparkpunkd) [inline packages/blueprints-integration/src/rundown.ts:303]: Is the comment supposed to be "... _undefined_ if not yet played back" or is the time `Time | null`?

File: packages/blueprints-integration/src/rundown.ts [hunk 2]
```diff
@@ -292,6 +299,13 @@ export interface IBlueprintPieceInstance<TMetadata = unknown> {
...
+	/** Whether the piece has stopped playback (the most recent time it was played).
+	 * This is set from a callback from the playout gateway (milliseconds since epoch)
+	 */
+	stoppedPlayback?: Time
```
Comment (sparkpunkd) [inline packages/blueprints-integration/src/rundown.ts:307]: Similarly, what is the not yet set value?

File: packages/blueprints-integration/src/rundown.ts (hunk 1 above)
Comment (sparkpunkd) [inline packages/blueprints-integration/src/rundown.ts:303]: Is the comment "... _undefined_ if not yet played back ... " or the type `Time | null` ?

File: meteor/server/api/blueprints/events.ts [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+				asyncCollectionFindOne(
+					PartInstances,
+					{
+						rundownId: partInstance.rundownId,
+						playlistActivationId: partInstance.playlistActivationId,
+						takeCount: { $lt: partInstance.takeCount },
+					},
+					{
+						sort: {
+							// TODO - test
```
Comment (sparkpunkd) [inline meteor/server/api/blueprints/events.ts:78]: Test done?

File: meteor/server/api/blueprints/context/context.ts [hunk 4]
```diff
@@ -594,91 +583,109 @@ export class AsRunEventContext extends RundownContext implements IAsRunEventCont
...
-				})
+	getFirstPartInstanceInRundown(): Readonly<IBlueprintPartInstance<unknown>> {
+		const partInstance = PartInstances.findOne(
+			{
+				rundownId: this._rundown._id,
+				playlistActivationId: this._part.playlistActivationId,
+			},
+			{
+				sort: {
+					// TODO - verify sort
```
Comment (sparkpunkd) [inline meteor/server/api/blueprints/context/context.ts:630]: Sort verified?

File: meteor/server/api/blueprints/context/context.ts [hunk 5]
```diff
@@ -545,41 +553,22 @@ export class TimelineEventContext extends RundownContext implements ITimelineEve
 	}
 }
 
-export class AsRunEventContext extends RundownContext implements IAsRunEventContext {
-	public readonly asRunEvent: Readonly<IBlueprintAsRunLogEvent>
-
+export class AsRunRundownEventContext extends RundownContext implements IAsRunRundownEventContext {
```
Comment (sparkpunkd) [inline meteor/server/api/blueprints/context/context.ts:556]: Semantically, is this still _AsRun..._ - I'm fine with this conceptually - this is the as run timing information for this rundown or part. Its just with the collection going away ...

**Review by sparkpunkd — APPROVED** [review]
Approved subject to some minor issues to address. Otherwise, time to say goodbye to unecesarry AsRun logs.

File: packages/blueprints-integration/src/rundown.ts (hunk 1 above)
Comment (Julusian) [reply packages/blueprints-integration/src/rundown.ts:303]: should read as undefined. Thats what I get for moving a property and comments without reading it..

File: meteor/server/api/blueprints/context/context.ts (hunk 5 above)
Comment (Julusian) [reply meteor/server/api/blueprints/context/context.ts]: Ive renamed these to be a bit clearer now

File: meteor/server/migration/0_25_0.ts [hunk 6]
```diff
@@ -98,18 +97,18 @@ export const addSteps = addMigrationSteps('0.25.0', [
...
-		'AsRunLog',
-		{
-			rundownId: 'runningOrderId',
-			// segmentId:	'segmentId',
-			partInstanceId: 'segmentLineId',
-			pieceInstanceId: 'segmentLineItemId',
-		},
-		'migrateDatabaseCollections'
-	),
+	// renamePropertiesInCollection(
```
Comment (nytamin) [inline meteor/server/migration/0_25_0.ts:100]: I'm missing a migration in x_x_x.ts which clears up the no-longer-used asRunLog collection. See how we did it in R25: https://github.com/nrkno/tv-automation-server-core/blob/master/meteor/server/migration/deprecatedDatabases/0_25_0.ts#L18

File: meteor/lib/lib.ts [hunk 7]
```diff
@@ -946,6 +946,9 @@ export function unprotectString(protectedStr: ProtectedString<any> | undefined |
 export function unprotectStringArray(protectedStrs: Array<ProtectedString<any>>): string[] {
 	return (protectedStrs as any) as string[]
 }
+export function unDeepString<T extends ProtectedString<any>>(str: ReadonlyDeep<T> | PartialDeep<T>): T {
```
Comment (nytamin) [inline meteor/lib/lib.ts:949]: should be unprotectStringDeep?

File: meteor/server/api/blueprints/context/context.ts [hunk 8]
```diff
@@ -594,91 +583,109 @@ export class AsRunEventContext extends RundownContext implements IAsRunEventCont
...
-	getParts(): Array<IBlueprintPartDB> {
-		return unprotectObjectArray(this._rundown.getParts())
+}
+
+export class RundownTimingEventContext extends RundownDataChangedEventContext implements IRundownTimingEventContext {
+	readonly previousPart: Readonly<IBlueprintPartInstance<unknown>> | undefined
+	private readonly _part: PartInstance
+	readonly nextPart: Readonly<IBlueprintPartInstance<unknown>> | undefined
+
+	public get part(): Readonly<IBlueprintPartInstance<unknown>> {
```
Comment (nytamin) [inline meteor/server/api/blueprints/context/context.ts:602]: Question: Should this be named currentPart, to be constistent with previous & next?

File: meteor/server/api/blueprints/context/context.ts [hunk 9]
```diff
@@ -594,91 +583,109 @@ export class AsRunEventContext extends RundownContext implements IAsRunEventCont
...
-		return []
+
+		return unprotectPartInstance(partInstance)
 	}
 
-	formatDateAsTimecode(time: number): string {
-		check(time, Number)
-		return formatDateAsTimecode(new Date(time))
+	getPartInstancesInSegmentPlayoutId(
+		refPartInstance: Readonly<IBlueprintPartInstance<unknown>>
```
Comment (nytamin) [inline meteor/server/api/blueprints/context/context.ts:647]: needs doc: what should the refPartInstance be?

Comment (nytamin) [reply meteor/server/api/blueprints/context/context.ts:647]: never mind, I found the answer in blueprints-integration

**Review by nytamin — CHANGES_REQUESTED** [review]
A few comments, looks good in general.

File: meteor/server/api/blueprints/context/context.ts (hunk 8 above)
Comment (Julusian) [reply meteor/server/api/blueprints/context/context.ts:602]: I did wonder that, but this is consistent with some other interfaces here.
But yeah I think I will change this to currentPart for better symmetry within this interface

File: meteor/lib/lib.ts (hunk 7 above)
Comment (Julusian) [reply meteor/lib/lib.ts:949]: should it? its converting `ReadonlyDeep<RundownId>` to `RundownId`, so its still a protected string

Comment (Julusian) [reply meteor/lib/lib.ts:949]: `unDeepProtectedString`?

**Pull request merged**